File: Leetcode/0023 Merge k Sorted Lists.py
class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        min_values = list(map(ComparableListNode, filter(bool, lists)))
        heapq.heapify(min_values)
        
        head = ListNode()
        tail = head
        while min_values:
            node = heapq.heappop(min_values).list_node
            tail.next = node
            tail = node
            if node.next:
                heapq.heappush(min_values, ComparableListNode(node.next))

        return head.next


class ComparableListNode:
    def __init__(self, list_node):
        self.list_node = list_node
    
    def __lt__(self, other):
        return self.list_node.val < other.list_node.val

File: Leetcode/0023 Merge k Sorted Lists.java
class Solution {
    private static final Comparator<ListNode> compareListNode = Comparator.comparingInt(node -> node.val);
    
    public ListNode mergeKLists(ListNode[] lists) {
        var minValues = new PriorityQueue<>(compareListNode);
        for(var head : lists) {
            if(head != null) {
                minValues.add(head);
            }
        }
        ListNode head = new ListNode(0), tail = head;
        while(!minValues.isEmpty()) {
            var node = minValues.poll();
            tail.next = node;
            tail = node;
            if(node.next != null) {
                minValues.add(node.next);
            }
        }
        return head.next;
    }
}
